src/client/src/Client.c

```diff
@@ -20,6 +20,84 @@ extern UINT32 CLIENT_STATE_MACHINE_STATE_COUNT;
  */
 PKinesisVideoClient gKinesisVideoClient = NULL;
 
+
+/**
+ *
+ * @param timerId - timerId for timer
+ * @param currentTime - the current time when the call back was fired
+ * @param customData - pKinesisVideoClient, contains the streams and interval info
+ * @return
+ */
+STATUS checkIntermittentProducerCallback(UINT32 timerId, UINT64 currentTime, UINT64 customData)
+{
+    ENTERS();
+    UNUSED_PARAM(timerId);
+    STATUS retStatus = STATUS_SUCCESS;
+    PKinesisVideoClient pKinesisVideoClient = (PKinesisVideoClient) customData;
+    UINT32 i;
+    BOOL clientStreamsListLocked = FALSE, streamLocked = FALSE;
+    PKinesisVideoStream pCurrStream = NULL;
+    Frame eofr = EOFR_FRAME_INITIALIZER;
+
+    CHK(pKinesisVideoClient, STATUS_NULL_ARG);
+
+    pKinesisVideoClient->timerCallbackInvocationCount++;
+
+    // Lock the client streams list lock
+    pKinesisVideoClient->clientCallbacks.lockMutexFn(pKinesisVideoClient->clientCallbacks.customData,
+                                                     pKinesisVideoClient->base.streamListLock);
+    clientStreamsListLocked = TRUE;
+    for (i = 0; i < pKinesisVideoClient->deviceInfo.streamCount; i++) {
+        if (NULL != pKinesisVideoClient->streams[i]) {
+            pCurrStream = pKinesisVideoClient->streams[i];
+            // Validate stream isn't being shut down or is closed
+            if (!(pCurrStream->streamStopped || pCurrStream->streamClosed)) {
+                // Lock the Stream
+                pKinesisVideoClient->clientCallbacks.lockMutexFn(pKinesisVideoClient->clientCallbacks.customData,
+                                                                 pCurrStream->base.lock);
+                streamLocked = TRUE;
+                // Check if last PutFrame is older than max timeout, if so, send EoFR, if not, do nothing
+                // Ignoring currentTime it COULD be smaller than pCurrStream->lastPutFrametimestamp
+                // Due to this method entering but waiting on stream lock from putFrame call
+                if (IS_VALID_TIMESTAMP(pCurrStream->lastPutFrameTimestamp) &&
+                    (GETTIME() - pCurrStream->lastPutFrameTimestamp) > INTERMITTENT_PRODUCER_MAX_TIMEOUT) {
+                    CHK_STATUS(putKinesisVideoFrame(TO_STREAM_HANDLE(pCurrStream), &eofr));
+                    pCurrStream->lastPutFrameTimestamp = INVALID_TIMESTAMP_VALUE;
+                }
+
+                // Unlock the Stream
+                pKinesisVideoClient->clientCallbacks.unlockMutexFn(pKinesisVideoClient->clientCallbacks.customData,
+                                                                   pCurrStream->base.lock);
+                streamLocked = FALSE;
+            }
+        }
+    }
+
+    if(clientStreamsListLocked) {

```

Comment:
spacing is off

Reply:
Also, no need to check for clientStreamListLocked here - it always is

Reply:
For consistency I kind of like the idea to just check everywhere, otherwise it's hard to spot where it's actually missing and should be there.  I suppose though if it occurs *before* the CleanUp section and the original lock is at the top of function (before any conditionals) then it will always be the case if we reached that part of execution we have the lock.  It looks like if I removed the CHK_STATUS there's no case where we can be holding onto a lock and jump to cleanup so I can in fact remove the unlock statements entirely from CleanUp, will do that.

Reply:
I like the fact that you are explicitly releasing the mutex in the "happy path". My issue is with checking whether it's locked. Static code analysis tools will flag this as an invariant. Just remove the if check and simply unlock the mutex and set the locked = FALSE;